pkg/apiserver/rest/dualwriter_syncer.go

```diff
@@ -0,0 +1,285 @@
+	syncSuccess := 0
+	syncErr := 0
+
+	maxInterval := dataSyncerInterval + 5*time.Minute
+
+	var errSync error
+	const maxRecordsSync = 1000
+
+	// LockExecuteAndRelease ensures that just a single Grafana server acquires a lock at a time
+	// The parameter 'maxInterval' is a timeout safeguard, if the LastExecution in the
+	// database is older than maxInterval, we will assume the lock as timeouted. The 'maxInterval' parameter should be so long
+	// that is impossible for 2 processes to run at the same time.
+	err := serverLockService.LockExecuteAndRelease(ctx, fmt.Sprintf("legacyToUnifiedStorageDataSyncer-%d-%s", mode, resource), maxInterval, func(context.Context) {
+		log.Info("starting legacyToUnifiedStorageDataSyncer")
+		startSync := time.Now()
+
+		orgId := int64(1)
+
+		ctx = klog.NewContext(ctx, log)
+		ctx = identity.WithRequester(ctx, getSyncRequester(orgId))
+		ctx = request.WithNamespace(ctx, requestInfo.Namespace)
+		ctx = request.WithRequestInfo(ctx, requestInfo)
+
+		storageList, err := getList(ctx, storage, &metainternalversion.ListOptions{
+			Limit: maxRecordsSync,
+		})
+		if err != nil {
+			log.Error(err, "unable to extract list from storage")
+			return
+		}
+
+		if len(storageList) >= maxRecordsSync {
+			errSync = fmt.Errorf("unified storage has more than %d records. Aborting sync", maxRecordsSync)
+			log.Error(errSync, "Unified storage has more records to be synced than allowed")
+			return
+		}
+
+		log.Info("got items from unified storage", "items", len(storageList))
+
+		legacyList, err := getList(ctx, legacy, &metainternalversion.ListOptions{})
+		if err != nil {
+			log.Error(err, "unable to extract list from legacy storage")
+			return
+		}
+		log.Info("got items from legacy storage", "items", len(legacyList))
+
+		itemsByName := map[string]syncItem{}
+		for _, obj := range legacyList {
+			accessor, err := utils.MetaAccessor(obj)
+			if err != nil {
+				log.Error(err, "error retrieving accessor data for object from legacy storage")
+				continue
+			}
+			name := accessor.GetName()
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
+			item.name = name
+			item.objLegacy = obj
+			itemsByName[name] = item
+		}
+
+		for _, obj := range storageList {
+			accessor, err := utils.MetaAccessor(obj)
+			if err != nil {
+				log.Error(err, "error retrieving accessor data for object from storage")
+				continue
+			}
+			name := accessor.GetName()
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
+			item.name = name
+			item.objStorage = obj
+			itemsByName[name] = item
+		}
+		log.Info("got list of items to be synced", "items", len(itemsByName))
+
+		for name, item := range itemsByName {
+			// upsert if:
+			// - existing in both legacy and storage, but objects are different, or
+			// - if it's missing from storage
+			if item.objLegacy != nil &&
+				((item.objStorage != nil && !Compare(item.objLegacy, item.objStorage)) || (item.objStorage == nil)) {
```

Comment:
[non-blocking] Right-hand of boolean `or` expression is only evaluated if left-hand is false.
```suggestion
				(item.objStorage == nil || !Compare(item.objLegacy, item.objStorage)) {
```